extension/httpforwarder/extension.go

```diff
@@ -16,21 +16,115 @@ package httpforwarder
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
 
 	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
 )
 
 type httpForwarder struct {
+	forwardTo  *url.URL
+	httpClient *http.Client
+	server     *http.Server
+	logger     *zap.Logger
+	config     *Config
 }
 
 var _ component.ServiceExtension = (*httpForwarder)(nil)
 
-func (h httpForwarder) Start(_ context.Context, _ component.Host) error {
-	// TODO: Start HTTP server
+func (h *httpForwarder) Start(_ context.Context, host component.Host) error {
+	listener, err := h.config.Ingress.ToListener()
+	if err != nil {
+		return fmt.Errorf("failed to bind to address %s: %w", h.config.Ingress.Endpoint, err)
+	}
+
+	handler := http.NewServeMux()
+	handler.HandleFunc("/", h.forwardRequests)
+
+	h.server = h.config.Ingress.ToServer(handler)
+	go func() {
+		if err := h.server.Serve(listener); err != nil {
+			host.ReportFatalError(err)
+		}
+	}()
+
 	return nil
 }
 
-func (h httpForwarder) Shutdown(_ context.Context) error {
-	// TODO: Shutdown HTTP server
-	return nil
+func (h *httpForwarder) Shutdown(_ context.Context) error {
+	return h.server.Close()
+}
+
+func (h *httpForwarder) forwardRequests(writer http.ResponseWriter, request *http.Request) {
+	forwarderRequest := request.Clone(request.Context())
+	forwarderRequest.URL.Host = h.forwardTo.Host
+	forwarderRequest.URL.Scheme = h.forwardTo.Scheme
+	forwarderRequest.Host = h.forwardTo.Host
+	// Clear RequestURI to avoid getting "http: Request.RequestURI can't be set in client requests" error.
+	forwarderRequest.RequestURI = ""
+
+	// Add additional headers.
+	for k, v := range h.config.Egress.Headers {
+		forwarderRequest.Header.Add(k, v)
+	}
+
+	// Add "Via" header for tracking purposes on both the outgoing requests and responses.
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Via.
+	addViaHeader(forwarderRequest.Header, request.Proto, request.Host)
+
+	response, err := h.httpClient.Do(forwarderRequest)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusBadGateway)
+		return
```

Comment:
is this guaranteed not to leak response resources (is response always nil)?

Reply:
I believe the response could be non-nil. Under normal circumstances as per the docs looks like it is safe to ignore the response when the error is non-nil ([golang doc](https://github.com/golang/go/blob/master/src/net/http/client.go#L566)). That said, in this case I guess the response should be relayed back to the caller. Updated to do so.